kafka/consumer.go

```diff
@@ -0,0 +1,393 @@
+					continue
+				}
+
+				c.conf.MessageHandler(e.Value, tp.Partition)
+				atomic.StoreInt64(offsetPtr, int64(tp.Offset))
+			case *confluent.Error:
+				log.Error(3, "kafka-consumer: Kafka consumer error: %s", e.String())
+				return
+			}
+		case <-c.stopChan:
+			log.Info("kafka-consumer: Consumer ended.")
+			return
+		}
+	}
+}
+
+func (c *Consumer) monitorLag(processBacklog *sync.WaitGroup) {
+	c.wg.Add(1)
+	defer c.wg.Done()
+
+	completed := make(map[int32]bool, len(c.Partitions))
+	for _, partition := range c.Partitions {
+		completed[partition] = false
+	}
+
+	storeOffsets := func(ts time.Time) {
+		for partition := range c.currentOffsets {
+			offset := atomic.LoadInt64(c.currentOffsets[partition])
+			c.partitionOffset[partition].Set(int(offset))
+			if c.LagMonitor != nil {
+				c.LagMonitor.StoreOffset(partition, offset, ts)
+			}
+			if !completed[partition] && offset >= c.bootTimeOffsets[partition]-1 {
+				if processBacklog != nil {
+					processBacklog.Done()
+				}
+				completed[partition] = true
+				delete(c.bootTimeOffsets, partition)
+				if len(c.bootTimeOffsets) == 0 {
+					c.bootTimeOffsets = nil
+				}
+			}
+
+			_, newest, err := c.consumer.QueryWatermarkOffsets(c.conf.Topics[0], partition, c.conf.MetadataTimeout)
+			if err != nil {
+				log.Error(3, "kafka-consumer: Error when querying for offsets: %s", err)
+			} else {
+				c.partitionLogSize[partition].Set(int(newest))
+			}
+
+			c.partitionOffset[partition].Set(int(offset))
+			if err == nil {
+				lag := int(newest - offset)
+				c.partitionLag[partition].Set(lag)
+				if c.LagMonitor != nil {
+					c.LagMonitor.StoreLag(partition, lag)
+				}
+			}
+		}
+	}
+
+	ticker := time.NewTicker(c.conf.OffsetCommitInterval)
+	for {
+		select {
+		case ts := <-ticker.C:
+			storeOffsets(ts)
+		case <-c.stopChan:
+			storeOffsets(time.Now())
+			return
+		}
+	}
+}
+
+func (c *Consumer) startConsumer() error {
+	var offset confluent.Offset
+	var err error
+	var topicPartitions confluent.TopicPartitions
+	c.currentOffsets = make(map[int32]*int64, len(c.Partitions))
+
+	for i, topic := range c.conf.Topics {
+		for _, partition := range c.Partitions {
+			var currentOffset int64
+			switch c.conf.StartAtOffset {
+			case "oldest":
+				currentOffset, _, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetBeginning), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			case "newest":
+				_, currentOffset, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetEnd), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			default:
+				offsetDuration, err := time.ParseDuration(c.conf.StartAtOffset)
+				if err != nil {
+					return fmt.Errorf("invalid offest format %s: %s", c.conf.StartAtOffset, err)
+				}
+				currentOffset = time.Now().Add(-1*offsetDuration).UnixNano() / int64(time.Millisecond)
+				currentOffset, _, err = c.tryGetOffset(topic, partition, currentOffset, 3, time.Second)
+				if err != nil {
+					return err
```

Comment:
we need to fall back to oldest, not return an error.

Reply:
actually this comment sounds like in this case it would just use oldest:
https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/consumer.go#L488-L490

but i'll fall back to oldest in case of any error now